syntax = "proto3";
package ipc;
option go_package = "actors/proto/ipc";

import "platform.proto";

enum RouterCommand {
  ROUTER_UNSPECIFIED = 0;
  ROUTER_ADD_HANDLER = 1;
  ROUTER_REMOVE_HANDLER = 2;
  ROUTER_EXECUTE = 3;
  ROUTER_EXIT = 4;
}

message AddHandler {
  string Name = 1;
  bytes Handler = 2;
  proto.Language Language = 3;
  repeated string Methods = 4;
}

message RemoveHandler {
  string Name = 1;
}

message Execute {
  uint64 ID = 1;
  string Name = 2;
  string Method = 3;
  map<string, proto.EncodedObject> Args = 4;
}

message Exit {}

message RouterMessage {
  RouterCommand Command = 1;
  oneof Payload {
    AddHandler AddHandler = 2;
    RemoveHandler RemoveHandler = 3;
    Execute Execute = 4;
    Exit Exit = 5;
  }
}
